### PR TITLE
[WIP] - Support Microsoft Edge (Chromium Based)

### LINKED ISSUE #5433: Support Microsoft Edge
### Current behavior:

Microsoft Edge is not supported.

### Desired behavior:

Microsoft Edge is supported. Although this issue is specifically about Anaheim (v74+), support for Spartan (v20-v44) would also be welcome. Given that Anaheim is pretty much the same as Chrome, this shouldn't be a major change, but it would allow Windows users to not have to install Chrome just for Cypress.

### Steps to reproduce: (app code and test code)

N/A

### Versions

Microsoft Edge 74+

### PR DESCRIPTION
**Pull request by YOU54F**
## Proposal for Microsoft Edge (Chromium Based) Support in Cypress

- close #5433

Offical Microsoft Edge Insiders page [here](https://www.microsoftedgeinsider.com/en-us/)

Currently tested

- [x] Microsoft Edge for Mac (Canary Build) 
  - Binary package [here](https://officecdn.microsoft.com/pr/C1297A47-86C4-4C1F-97FA-950631F94777/MacAutoupdate/MicrosoftEdgeCanary-76.0.151.0.pkg)
- [x] Microsoft Edge for Mac (Dev Build) 
  - Binary package [here](https://officecdn.microsoft.com/pr/C1297A47-86C4-4C1F-97FA-950631F94777/MacAutoupdate/MicrosoftEdgeDev-76.0.151.0.pkg)
- [X] Microsoft Edge for Windows 10 (Canary Build)
  - Binary package [here](https://c2rsetup.officeapps.live.com/c2r/downloadEdge.aspx?ProductreleaseID=Edge&platform=Default&version=Edge&Channel=Canary&language=en-us&Consent=1&IID=fe1b2a91-d63d-532d-ac23-eaad1e276343)
- [X] Microsoft Edge for Windows 10 (Dev Build) 
  - Binary package [here](https://c2rsetup.officeapps.live.com/c2r/downloadEdge.aspx?ProductreleaseID=Edge&platform=Default&version=Edge&Channel=Dev&language=en-us&Consent=0&IID=fe1b2a91-d63d-532d-ac23-eaad1e276343)
- [ ] Microsoft Edge for Linux (Canary Build)
  - Binary package currently unavalable
- [ ] Microsoft Edge for Linux (Dev Build)
  - Binary package currently unavalable
  
Terminal output after change (in `/packages/launcher`)

```sh
➜  launcher git:(feat/msEdge) ✗ npm run build && node index.js
Alias tip: npmR build && npm install && node index.js

> @packages/launcher@0.0.0 build /Users/you54f/dev/saftest/githubrepos/forks/cypress/packages/launcher
> bin-up tsc --project .

Found /Users/you54f/dev/saftest/githubrepos/forks/cypress/node_modules/.bin/tsc

Launcher project exports
{ launch: [Function: launch],
  detect: [Function],
  detectByPath: [Function] }
⛔️ please use it as a module, not from CLI
detected 2 browsers
[ { name: 'chrome',
    family: 'chrome',
    displayName: 'Chrome',
    version: '74.0.3729.157',
    path:
     '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
    majorVersion: '74' },
  { name: 'edge',
    family: 'chrome',
    displayName: 'Edge',
    version: '76.0.159.0',
    path:
     '/Applications/Microsoft Edge Canary.app/Contents/MacOS/Microsoft Edge Canary',
    majorVersion: '76' } ]
```
Screenshot 2019-05-14 at 22 59 23
Screenshot 2019-05-14 at 22 59 28
Screenshot 2019-05-14 at 23 11 03
Screenshot 2019-05-14 at 23 26 02
Screenshot 2019-05-14 at 23 26 12

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "cypress",
+  "name": "@you54f/cypress",
```
Comment (jennifer-shehane) [inline package.json]: We'll be unable to accept these changes to the package.json name and url. Do you have some autoupdate on your precommit or IDE that changes this?

Comment (YOU54F) [reply package.json]: Hey, totally understand and  sorry didn’t mean to add those changes to this PR, for the url and package name. They were changed so I could test the changes with a packaged product  🙂

I don’t think you will want to accept the changes in the pr until new edge moves from beta release, so I think this pr might be a bit of a a slow burner

Comment (YOU54F) [reply package.json]: Hey, I've now removed the erroneous `package.json` changes 👍 thanks for updating the title too.

**Comment by YOU54F** [comment]
@jennifer-shehane thanks for reviewing, I am tracking and testing the edge packages at this github repo https://github.com/YOU54F/cypress-edge and have it working in CI at the moment :)

**Comment by coding-yogi** [comment]
Need additional changes to make it work on windows

**Comment by YOU54F** [comment]
> Need additional changes to make it work on windows

🙄 yep, hence why I’ve not marked the windows checkboxes in the first post.

**Comment by YOU54F** [comment]
now supporting windows edge/dev ms edge builds

```
detected 3 browsers
[ { name: 'chrome',
    family: 'chrome',
    displayName: 'Chrome',
    version: '74.0.3729.169',
    path:
     'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
    majorVersion: '74' },
  { name: 'edgeCanary',
    family: 'chrome',
    displayName: 'Edge Canary',
    version: '77.0.186.0',
    path:
     'C:\\Users\\grung\\AppData\\Local\\Microsoft\\Edge SxS\\Application\\msedge.exe',
    majorVersion: '77' },
  { name: 'edgeDev',
    family: 'chrome',
    displayName: 'Edge Dev',
    version: '76.0.176.1',
    path:
     'C:\\Program Files (x86)\\Microsoft\\Edge Dev\\Application\\msedge.exe',
    majorVersion: '76' } ]
```

**Comment by YOU54F** [comment]
@flotwig @bahmutov 

Got Edge Dev & Edge Canary running on Windows now but I am having trouble building the binary.

It gets most of the way but fails when trying to perform a symlink for the electron package.

I am running git bash for windows (https://gitforwindows.org/)

```
All tasks completed successfully
1:07:54 AM #testVersion win32
testing dist package version
by calling: node index.js --version
in the folder C:\Users\grung\Documents\GitHub\cypress\dist\win32
app in C:\Users\grung\Documents\GitHub\cypress\dist\win32
built app version 3.3.0
1:07:56 AM #elBuilder win32
from C:\Users\grung\Documents\GitHub\cypress\dist\win32
into C:\Users\grung\Documents\GitHub\cypress\build\win32\Cypress
Packaging app for platform win32 x64 using electron v2.0.18
Error: EPERM: operation not permitted, symlink 'C:\Users\grung\AppData\Roaming\Cypress\cy\' -> 'C:\Users\grung\AppData\Local\Temp\electron-packager\win32-x64\Cypress-win32-x64\resources\app\packages\server\.cy'
```

How are you building it for windows? Via Appveyor?

BTW - Your contributing and deployment docs are quality! and the build process is 👌 kudos to all in the team.

**Comment by flotwig** [comment]
@YOU54F Hmm, I don't know if I've seen that, but re-building Cypress on Windows from scratch always fixes whatever weird filesystem issue I'm having. Have you tried re-cloning Cypress and doing `npm install` over again? Might fix it.

**Comment by YOU54F** [comment]
> @YOU54F Hmm, I don't know if I've seen that, but re-building Cypress on Windows from scratch always fixes whatever weird filesystem issue I'm having. Have you tried re-cloning Cypress and doing `npm install` over again? Might fix it.

Cheers for the reply bud, I'll try it on my desktop machine when I'm home, I've got a few diff shells on that, and WSL (windows subsystem linux) so may have more luck. Will keep you posted 👍

**Comment by bahmutov** [comment]
@YOU54F I have confirmed it is working on Mac, could you accept https://github.com/YOU54F/cypress/pull/5 that fixes the linting errors? For some reason my CLI pull request never upload to this PR, instead they create a new branch, ughh

**Comment by bahmutov** [comment]
ughh, Edge is not available on our https://www.appveyor.com/docs/windows-images-software/#web-browsers

**Comment by YOU54F** [comment]
@bahmutov accepted the pr now for linter errors.

Chromium edge is still only available on insiders at the moment, so you would need to download and install the executable directly on a windows vm. I have been meaning to set it up with circleci now that they have windows executors

**Comment by bahmutov** [comment]
yeah @YOU54F - I would love to see how it is done on Windows.
Also, this PR is interesting from https://github.com/cypress-io/cypress/pull/5068 - since if we can return MS Edge in the list of browsers with `family: chrome` there it will "just" work.

**Comment by YOU54F** [comment]
> yeah @YOU54F - I would love to see how it is done on Windows.
> Also, this PR is interesting from #5068 - since if we can return MS Edge in the list of browsers with `family: chrome` there it will "just" work.

That PR does indeed look like it would be ideal! I think there was someone on gitter trying to get chrome to recognise a browser that chrome based but didn't mention which one, so it looks like it would get many birds with one stone.

Will see what I can do with regards to the windows edge on circleci with cypress 👍 got the laptop out as I have been doing some tweaks on a work project to see if we can speed up build times but that can really wait until the 9-5!

**Comment by bahmutov** [comment]
yeah, not rush there. I think this PR is great because it has the detection logic.

**Comment by YOU54F** [comment]
Ahhh reading my last comments, I forgot that I couldn't actually get it to build the binaries on my windows machines!

**Comment by YOU54F** [comment]
@bahmutov - tried to setup a run in circleci, but I've ran out of credits so no dice, but got the binaries built in windows (via bootcamp on my mac) and have it working now.

I have released the binaries here :- https://github.com/YOU54F/cypress/releases/download/v3.5.0/cypress_win.zip

and an associated cli release here :- https://www.npmjs.com/package/@you54f/cypress

instructions to use, for testing

##  Supported versions
- [X] Microsoft Edge for Windows 10 (Canary Build)
  - Binary package [here](https://c2rsetup.officeapps.live.com/c2r/downloadEdge.aspx?ProductreleaseID=Edge&platform=Default&version=Edge&Channel=Canary&language=en-us&Consent=1&IID=fe1b2a91-d63d-532d-ac23-eaad1e276343)
- [X] Microsoft Edge for Windows 10 (Dev Build) 
  - Binary package [here](https://c2rsetup.officeapps.live.com/c2r/downloadEdge.aspx?ProductreleaseID=Edge&platform=Default&version=Edge&Channel=Dev&language=en-us&Consent=0&IID=fe1b2a91-d63d-532d-ac23-eaad1e276343)

## Instructions for Windows
1. Download Microsoft Edge version of choice as above
2. Make a new directory
3. Run `set CYPRESS_INSTALL_BINARY=https://github.com/YOU54F/cypress/releases/download/v3.5.0/cypress_win.zip`
4. Run `npm init`
5. Run `npm install @you54f/cypress --save`
6. Run `node_modules/.bin/cypress open --browser edgeDev` to open in interactive mode, and setup [Cypress.io](https://cypress.io)'s example project
7. Run `node_modules/.bin/cypress run --browser edgeDev` or `node_modules/.bin/cypress run --browser edgeCanary` to run in command line mode.
8. Rejoice & please pass back some appreciation with a star on the [repository](https://github.com/YOU54F/cypress-edge)! Thanks :)

**Comment by jennifer-shehane** [comment]
Hey @YOU54F, any updates on this PR?

**Comment by YOU54F** [comment]
> Hey @YOU54F, any updates on this PR?

Hello @jennifer-shehane,

Edge is due out on general release near the end of the January, it was announced near the start or December. There is no news on a Linux build yet, so it will be macOS and windows only. 

We could release the Mac changes now and the windows changes when we have the official executable names when it goes on general release or support edge dev, beta and canary until the released versions are out?

**Comment by jennifer-shehane** [comment]
@YOU54F Ok, I just wanted to hear what you felt was left. There's a lot more left on our side:

- Updating our tests to run through Edge (this may show potential issues between Edge and our other tested browser that we'd need to fix)
- Adding an Edge icon to the browser dropdown
- Creating a docker container with Edge to use
- Add support within browser functions like 'isBrowser'
- Handled excluding Edge from Linux builds

These are just off the top of my head, there's a lot more that needs updating, you can see some in the Firefox PR that we would need to do for any added Browser https://github.com/cypress-io/cypress/pull/1359/files 

I suspect we'll have someone from our team start a new PR for this work is what I'm getting at. cc @bahmutov

**Comment by bahmutov** [comment]
The MS Edge went into wide availability https://blogs.windows.com/windowsexperience/2020/01/15/new-year-new-browser-the-new-microsoft-edge-is-out-of-preview-and-now-available-for-download/

**Comment by bahmutov** [comment]
@YOU54F thank you for this excellent work, I have pulled your branch into internal branch and opened a pull request https://github.com/cypress-io/cypress/pull/6181 - it seems to work on my Mac, but I need to set up some kind of CI test before landing this. Closing this PR, follow 6181 for updates

**Comment by YOU54F** [comment]
Cheers @bahmutov & @jennifer-shehane glad to help!

**Pull request closed without merging**